File: sorting-searching/sorting-algorithms/top-k-selection/TopKSelection.java
import java.util.Arrays;
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.List;
import java.util.ArrayList;

public class TopKSelection {

    public static List<Integer> topKLargest(int[] arr, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for (int num : arr) {
            if (minHeap.size() < k) {
                minHeap.add(num);
            }
            else if (num > minHeap.peek()) {
                minHeap.poll();
                minHeap.add(num);
            }
        }

        List<Integer> result = new ArrayList<>();

        while (!minHeap.isEmpty()) {
            result.add(minHeap.poll());
        }

        result.sort(Collections.reverseOrder());

        return result;
    }

    public static void main(String[] args) {
        int[] arr = {12, 5, 7, 19, 3, 25, 8};
        int k = 3;

        List<Integer> result = topKLargest(arr, k);

        System.out.println("Top " + k + " largest elements: " + result);
    }
}

File: sorting-searching/sorting-algorithms/top-k-selection/top-k-selection.py
import heapq

def top_k_largest(arr, k):
    min_heap = []

    for num in arr:
        if len(min_heap) < k:
            heapq.heappush(min_heap, num)

        elif num > min_heap[0]:
            heapq.heappop(min_heap)
            heapq.heappush(min_heap, num)

    result = sorted(min_heap, reverse=True)

    return result


arr = [12, 5, 7, 19, 3, 25, 8]
k = 3

result = top_k_largest(arr, k)

print(f"Top {k} largest elements:", result)
